packages/compartment-mapper/src/parse-cjs-shared-export-wrapper.js

```diff
@@ -0,0 +1,111 @@
+const { apply } = Reflect;
+const { freeze, keys, create, hasOwnProperty } = Object;
+
+/**
+ * @param {Object} object
+ * @param {string} key
+ * @returns {boolean}
+ */
+const has = (object, key) => apply(hasOwnProperty, object, [key]);
+
+/**
+ * ModuleEnvironmentRecord wrapper
+ * Creates shared export processing primitives to be used both Location and Archive usecases of cjs
+ *
+ * @param {Object} moduleEnvironmentRecord
+ * @param {Compartment} compartment
+ * @param {Record<string, string>} resolvedImports
+ * @returns {{
+ *   module: { exports: any },
+ *   moduleExports: any,
+ *   afterExecute: Function,
+ *   require: Function,
+ * }}
+ */
+export const wrap = (moduleEnvironmentRecord, compartment, resolvedImports) => {
+  // This initial default value makes things like exports.hasOwnProperty() work in cjs.
+  moduleEnvironmentRecord.default = create(
+    compartment.globalThis.Object.prototype,
+  );
+
+  // Set all exported properties on the defult and on the namespace simultaneously for import *
+  // Root namespace is only accessible for imports. Requiring from cjs gets the default field of the namespace
+  const assignProp = (prop, value) => {
+    // TODO Consider __esModule handling
+    if (prop === '__esModule') {
+      return;
+    }
+    if (prop !== 'default') {
+      moduleEnvironmentRecord[prop] = value;
+    }
+    moduleEnvironmentRecord.default[prop] = value;
+  };
+
+  const originalExports = new Proxy(Object.create(null), {
+    get(target, prop) {
+      return moduleEnvironmentRecord.default[prop];
+    },
+    set(_target, prop, value) {
+      assignProp(prop, value);
+      return true;
+    },
+    defineProperty(_target, prop, descriptor) {
+      // Object.defineProperty(moduleEnvironmentRecord, prop, descriptor) would work if someone in the npm
+      // community didn't come up with the idea to redefine property in the getter.
+      if (has(descriptor, 'value')) {
+        // This will result in non-enumerable properties being enumerable, but it's better than undefined
+        assignProp(prop, descriptor.value);
+      } else {
+        // Running the getter defeats some lazy initialization tricks.
+        // Whoever depends on getters on exports for lazy init instead
+        // of factory functions should face the consequences.
+        assignProp(prop, descriptor.get ? descriptor.get() : undefined);
+      }
+      return true;
+    },
+  });
+
+  let finalExports = originalExports;
+
+  const module = freeze({
+    get exports() {
+      return finalExports;
+    },
+    set exports(value) {
+      finalExports = value;
+    },
+  });
+
+  const require = freeze((/** @type {string} */ importSpecifier) => {
+    const namespace = compartment.importNow(resolvedImports[importSpecifier]);
+    // If you read this file carefully, you'll see it's not possible for a cjs module to not have the default anymore.
+    // It's currently possible to require modules that were not created by this file though.
```

Comment:
so now `require('a-package-that-has-esm-exports')` will get the default if present and the entire namespace if not. 

This may be controversial, but at this time it seems to be working great (at least for examples I've got in e2e tests)

If we want to fix it, we'd have to make `require` get the cjs variant of the package, parsed as such.

Reply:
This is fine. We could on the one hand create separate compartments for the CJS or ESM interpretation of a single package and thread them accordingly, but that would risk causing identity discontinuity between ESM and CJS instances.

This is an option we might present to developers, something they can express in an app’s package.json, but I’m content to avoid the problem for now.